cmd/sidecar-e2e/main.go

```diff
@@ -0,0 +1,353 @@
+		"-p", strconv.Itoa(dnsmasqPort),
+		"-c", "1337",
+		"-8", "-",
+		"-A", "/ok.local/1.2.3.4",
+		"-A", "/nxdomain.local/",
+	}
+	t.dnsmasq = exec.Command(opts.dnsmasqBinary, args...)
+
+	log.Printf("Starting dnsmasq %v", args)
+	startDumpToBuf(&t.lock, &t.dnsmasqOutput, t.dnsmasq)
+	err := t.dnsmasq.Start()
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	if err := waitForTcpOrExit(t.dnsmasq, fmt.Sprintf("127.0.0.1:%v", dnsmasqPort)); err != nil {
+		log.Fatal(err)
+	}
+	log.Printf("dnsmasq started")
+}
+
+func (t *test) runSidecar() {
+	args := []string{
+		"-v", "4",
+		"--prometheus-port", strconv.Itoa(sidecarPort),
+		"--dnsmasq-port", strconv.Itoa(dnsmasqPort),
+		"--probe",
+		fmt.Sprintf("ok,127.0.0.1:%v,ok.local,1", dnsmasqPort),
+		"--probe",
+		fmt.Sprintf("nxdomain,127.0.0.1:%v,nx.local,1", dnsmasqPort),
+		"--probe",
+		fmt.Sprintf("notpresent,127.0.0.1:%v,notpresent.local,1", dnsmasqPort+1),
+	}
+	t.sidecar = exec.Command(opts.sidecarBinary, args...)
+
+	log.Printf("Starting sidecar %v", args)
+	startDumpToBuf(&t.lock, &t.sidecarOutput, t.sidecar)
+	err := t.sidecar.Start()
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	if err := waitForTcpOrExit(t.sidecar, fmt.Sprintf("127.0.0.1:%v", sidecarPort)); err != nil {
+		log.Fatal(err)
+	}
+	log.Printf("sidecar started")
+}
+
+func (t *test) runDig() {
+	log.Printf("running `dig`")
+	for i := 0; i < 100; i++ {
+		dig := exec.Command(
+			opts.digBinary, "@127.0.0.1", "-p", strconv.Itoa(dnsmasqPort), "localhost")
+		output, err := dig.CombinedOutput()
+		if err == nil {
+			t.digOutput += string(output)
+		} else {
+			t.errors = append(t.errors, fmt.Errorf("Error running dig: %v", err))
+		}
+	}
+}
+
+func (t *test) waitForMetrics() {
+	log.Printf("Waiting for hits to be reported to be greater than 100")
+	waitForPredicate(
+		func() (bool, error) {
+			response, err := http.Get(
+				fmt.Sprintf("http://127.0.0.1:%v/metrics", sidecarPort))
+			if err != nil {
+				log.Fatal(err)
+			}
+
+			defer response.Body.Close()
+			buf, err := ioutil.ReadAll(response.Body)
+
+			lines := strings.Split(string(buf), "\n")
+			for _, line := range lines {
+				if !strings.HasPrefix(line, "kubedns_dnsmasq_hits ") {
+					continue
+				}
+
+				parts := strings.Split(line, " ")
+				if len(parts) < 2 {
+					return false, fmt.Errorf("invalid output for kubedns_dnsmasq_hits metric")
+				}
+
+				value, err := strconv.Atoi(parts[1])
+				if err != nil {
+					return false, fmt.Errorf("invalid output for kubedns_dnsmasq_hits metric")
+				}
+
+				if value >= 100 {
+					return true, nil
+				}
+			}
+			return false, nil
+		},
+		10*time.Second,
+		1*time.Second)
+}
+
+func (t *test) getMetrics() {
+	response, err := http.Get(fmt.Sprintf("http://127.0.0.1:%v/metrics", sidecarPort))
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	defer response.Body.Close()
+	buf, err := ioutil.ReadAll(response.Body)
+	if err != nil {
+		log.Fatal(err)
+	}
+	t.metricsOutput = string(buf)
+}
+
+func (t *test) dump() {
```

Comment:
println outputs to stdout which is captured by the harness, log uses stderr which is shown to the user